src/libraries/CalldataDecoder.sol

```diff
@@ -11,35 +11,61 @@ library CalldataDecoder {
+            // Strict encoding requires that the data begin with:
+            // 0x00: 0x40 (offset to `actions.length`)
+            // 0x20: 0x60 + actions.length (offset to `params.length`)
+            // 0x40: `actions.length`
+            // 0x60: beginning of actions
+
+            // Verify actions offset matches strict encoding
+            let invalidData := xor(calldataload(_bytes.offset), 0x40)
+            actions.offset := add(_bytes.offset, 0x60)
+            actions.length := and(calldataload(add(_bytes.offset, 0x40)), OFFSET_OR_LENGTH_MASK)
 
-            // The length is stored as the first element
-            actions.length := calldataload(actionsPtr)
-            params.length := calldataload(paramsPtr)
+            // Round actions length up to be word-aligned, and add 0x60 (for the first 3 words of encoding)
+            let paramsLengthOffset := add(and(add(actions.length, 0x1f), OFFSET_OR_LENGTH_MASK_AND_WORD_ALIGN), 0x60)
+            // Verify params offset matches strict encoding
+            invalidData := or(invalidData, xor(calldataload(add(_bytes.offset, 0x20)), paramsLengthOffset))
+            let paramsLengthPointer := add(_bytes.offset, paramsLengthOffset)
+            params.length := and(calldataload(paramsLengthPointer), OFFSET_OR_LENGTH_MASK)
+            params.offset := add(paramsLengthPointer, 0x20)
 
-            // The actual data is stored in the slot after the length
-            actions.offset := add(actionsPtr, 0x20)
-            params.offset := add(paramsPtr, 0x20)
+            // Expected offset for `params[0]` is params.length * 32
+            // As the first `params.length` slots are pointers to each of the array element lengths
+            let tailOffset := shl(5, params.length)
+            let expectedOffset := tailOffset
 
-            // Calculate how far `params` is into the provided bytes
-            let relativeOffset := sub(params.offset, _bytes.offset)
-            // Check that that isn't longer than the bytes themselves, or revert
-            if lt(_bytes.length, add(params.length, relativeOffset)) {
+            for { let offset := 0 } lt(offset, tailOffset) { offset := add(offset, 32) } {
+                let itemLengthOffset := calldataload(add(params.offset, offset))
+                // Verify that the offset matches the expected offset from strict encoding
+                invalidData := or(invalidData, xor(itemLengthOffset, expectedOffset))
+                let itemLengthPointer := add(params.offset, itemLengthOffset)
+                let length :=
+                    add(and(add(calldataload(itemLengthPointer), 0x1f), OFFSET_OR_LENGTH_MASK_AND_WORD_ALIGN), 0x20)
+                expectedOffset := add(expectedOffset, length)
+            }
+
+            // if the data encoding was invalid, or the provided bytes string isnt as long as the encoding says, revert
+            if or(invalidData, lt(add(_bytes.length, _bytes.offset), add(params.offset, expectedOffset))) {
```

Comment:
This check tolerates extra bytes in `_bytes` after encoded `params`.  Consider requiring strict equality between end of `_bytes` and end of encoded `params`.